src/journal/journald-context.c

```diff
@@ -268,6 +274,54 @@ static int client_context_read_label(
         return 0;
 }
 
+static int client_context_read_log_filter_xattr(const char *cgroup, Set **patterns, const char *key) {
+        _cleanup_free_ char *xattr = NULL;
+        int r;
+
+        r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, cgroup, key, &xattr);
+        if (r == -ENODATA) {
+                /* -ENODATA is returned when the key is not defined, this is not an error. */
+                *patterns = set_free_with_destructor(*patterns, pattern_free);
+                return 0;
+        }
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get %s xattr for %s: %m", key, cgroup);
+
+        const char *pattern;
+        NULSTR_FOREACH(pattern, xattr) {
+                pcre2_code *regex = NULL;
+
+                r = pattern_compile_and_log(pattern, 0, &regex);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to compile log filtering pattern '%s' for %s: %m",
+                                               pattern, cgroup);
+
+                r = set_ensure_put(patterns, NULL, regex);
+                if (r < 0) {
+                        regex = pattern_free(regex);
+                        return log_debug_errno(r, "Failed to insert regex into set for %s: %m", cgroup);
+                }
+        }
+
+        return 0;
+}
+
+static int client_context_read_log_filter_patterns(ClientContext *c, const char *cgroup) {
+        int r;
+
+        r = client_context_read_log_filter_xattr(cgroup, &c->log_filter_allowed_patterns,
+                                                 "user.journald_log_filter_allowed_patterns");
+        if (r < 0)
+                return r;
+
+        r = client_context_read_log_filter_xattr(cgroup, &c->log_filter_denied_patterns,
+                                                 "user.journald_log_filter_denied_patterns");
+        if (r < 0)
+                return r;
```

Comment:
hmm, So I am just realizing: there's a bit of a racy here: since these are two xattrs pid1 will write one first and the second after wards. If we race against that here we might end up reading the allow list from one invocation and the deny list from another, which might lead to temporary overfiltering or underfiltering, both of which might be a security issue.

Let's maybe merge the two xattrs: just encode both lists in the same nulstr after all, and mark the deny entries somehow (with the "~" prefix?) or insert some separator between the two prats, for example a 0xFF byte or so. (i find the latter a bit nicer)